### ISSUE TITLE
Signal exporting not working

### ISSUE THREAD

**Issue by penguinjuce311** [godotengine/godot#77455]
### Godot version

4.0.2.stable

### System information

windows 10

### Issue description

When you export a signal in a script and use it on a node a slot comes up on the editor to assign the value but nothing happens when its clicked

### Steps to reproduce

have a script with an export variable of type Signal

### Minimal reproduction project

n/a

**Comment by AThousandShips** [comment]
This is very vague and little information, what is the issue? What are you expecting?

**Comment by KoBeWi** [comment]
EditorPropertySignal seems non-functional. Pressing the button emits `object_id_selected`, but I couldn't find what it does. Also the signal name doesn't display even if you assign it explicitly.

I think this class is a stub. Exporting signals used to lead to a crash, now we have this fake property editor. I think the button should be disabled, signals are not even serializable AFAIK.

**Comment by Deozaan** [comment]
I'm trying to understand the use case. Under what circumstances would you want to export a signal?

**Comment by nkrisc** [comment]
> I'm trying to understand the use case. Under what circumstances would you want to export a signal?

Not the OP, but I ran into this as well. I have created a re-usable component scene that requires some signal for its logic, and I'd like to be able to connect it to an arbitrary signal in my scene, as depending on where I use it, it will require different signals.

Certainly there are many ways to work around this, so it's not a blocker of any kind, but it would have been nice to construct my entities in the editor with this component and make this signal connection without need to write extra code just for that. Currently I can build various entities from components without writing any additonal code.

**Comment by Deozaan** [comment]
So you don't want to export a signal, you want to export a variable that will reference an external signal that you want to connect to?

**Comment by nkrisc** [comment]
> So you don't want to export a signal, you want to export a variable that will reference an external signal that you want to connect to?

Is that not what this topic is about? What else could “exporting a signal” mean?

When I write:

```
@export var some_signal : Signal
```

It appears in the inspector but the button is non-functional.

Whereas 

```
@export signal my_signal
```

wouldn’t really make any sense at all, Side signals essentially are “exported” already in the Signal tab of the inspector.

**Comment by AThousandShips** [comment]
I'm not sure what the usefulness of this is, as signals can't and shouldn't be serialised, so what does editing them in the editor actually accomplish?

**Comment by KoBeWi** [comment]
The serialization can be implemented as a combination of NodePath (node) and StringName (signal name).

**Comment by AThousandShips** [comment]
Only partially, you can have a signal for a non-Node object

**Comment by Deozaan** [comment]
> What else could “exporting a signal” mean?

I really had no idea. 😅 I never thought of signals as "reference-able" from a variable before. I had always thought of signals as similar to constants. Something you know in advance and don't change, and therefore just access directly.

I asked for a use-case to try to get a more concrete explanation of what exporting a signal could possibly accomplish. It's a paradigm shift for me. So thank you for helping me understand.

**Comment by nkrisc** [comment]
> It's a paradigm shift for me. So thank you for helping me understand.

For me as well, I only realized I wanted this feature a day or two ago.

> Only partially, you can have a signal for a non-Node object

Perhaps that would be a sufficient compromise, though? Connecting signals through the inspector currently has a similar limitation, does it not? You can only connect signals from Nodes to methods on other Nodes using the inspector GUI. Even today if I want to connect to a signal from a Resource I need to do it through code. So it wouldn't be too unreasonable, in my opinon at least, for exporting a reference to a Signal to have the same limitation.

**Comment by AThousandShips** [comment]
I don't see this being a useful feature to have, it's better to handle signals in `_ready` or other places to handle the interactions, signals really aren't designed to work this way IMHO

I consider the proper solution to this to simply not handle signals and callables in the editor and restrict their exports, because it doesn't align with how their intended use is

**Comment by nkrisc** [comment]
> I don't see this being a useful feature to have, it's better to handle signals in _ready or other places to handle the interactions, signals really aren't designed to work this way IMHO

I would find it useful from a perspective of being able to make more modular scenes. I have a scene that has actions that are triggered by a signal, but there are many possible signals (from a variety of objects) I might want to use to trigger that action. 

Again, there are alternative approaches that work, so it's not a large issue. But for my use case it would be a very useful feature that reduces the complexity of my code and allows more configuration within the editor itself. I prefer to create workflows for myself that reduce or eliminate the need for additional code to be written and allow me to work in the editor as much as possible.

Two alternative approaches I've considered are: using the Autoload event bus pattern, or a class that takes a node reference and a StringName for the signal to act as a universal "plug" for the signal. Both have their drawbacks though, mostly revolving around requiring more boilerplate code to be written. Either I need to remember create new signals in my event bus or I need to add "signal sockets" to everything. Not a big problem, but not ideal in my opinion.

Anyway, that's all I have to add, so I'll leave it at that and keep my eye on this issue.

**Comment by Measurity** [comment]
> I would find it useful from a perspective of being able to make more modular scenes. I have a scene that has actions that are triggered by a signal, but there are many possible signals (from a variety of objects) I might want to use to trigger that action.

Yes, what you're describing is a common workflow for UI development. Like with the MVVM (Model-View-ViewModel) paradigm. For those interested you can read more about it here: https://docs.avaloniaui.net/docs/data-binding/binding-to-commands. This would be a nice-to-have for me.

**Comment by SwatDoge** [comment]
As an additional usecase, I have a state machine with a reusable `StateAwaitingSignal` node. This state simply waits for the selected  signal to emit, after which it transitions to the next state.

With this setup, I need to fill in both the Node and the name of its signal. It would be nicer if I could select the signal directly. This same logic can be used to connect UI to signals.

![image](https://github.com/godotengine/godot/assets/39221207/cee39b3e-559f-4c06-af1a-acd6c68adec7)

```GDScript
class_name GameplayLoopStateAwaiting
extends GameplayLoopState

## The id of this state. Used for transitions.
@export var state_id: String = ""
## Parent node of the signal that this state is waiting on.
@export var signal_element: Node
## The signal that this state is waiting on.
@export var signal_name: String = ""

@export_group("State after events")
## The state id that you will transition to after this state completes succesfully.
@export var signal_success: String = ""

func enter() -> void:
	if signal_element[signal_name]:
		await signal_element[signal_name]
		state_transitioned.emit(state_id, signal_success)
	else:
		printerr("Node or signal does not exist")
```

**Comment by MikeFP** [comment]
I just hit a limitation that's related to this issue.

I have setup a connection in the editor from nodes `A -> B`. However, those nodes act more as a template, and actually get duplicated during runtime whenever they need to be used.

Let's call the duplicated nodes  `A copy` and `B copy`. Using `Node.duplicate()` by code on `A`, the connections from `A copy -> B` persist, as expected. However, duplicating only the `B` node, incoming connections from `A -> B copy` don't persist.

Some work-arounds I thought of:

1. **Setup `A -> B` as a standalone scene**

Then, I could just `PackedScene.instantiate()`. However, that's not desirable for me. Those Nodes are already a part of a scene that has those connections. Making an extra scene just for that connection is just boilerplate, in my opinion, and will scale badly with the amount of those "template" nodes and connections between them. Those Nodes can function by themselves already, and shouldn't need to be put in scenes just for the sake of it. Also, `Node.duplicate()` should cover this usecase, but it doesn't.

2. **Copy incoming signal connections after duplicating**

I tried doing that by reconnecting the signals using the info from `Node.get_incoming_connections()`, and it worked almost perfectly, **except** for this: the "Advanced" options setup via the "Edit signal connection" window are lost.

That breaks, for instance, if the connection was setup to a method with a different signature by using the "Unbinds" option. Since that information is saved in the scene, the Engine is able to connect those signals when the scene is instantiated, but the node itself doesn't have access to those advanced options in order to recreate the connection.

Therefore, this approach is way less powerful than connections setup via the Editor, and it won't work for every connection.

3. **Connect to the signal at runtime using exported variables**

This would circumvent the need to duplicate the signal connections, since they would always be setup in `_ready()`. However, I'd still lose the ability to connect to methods of different signatures (see point 2), which is only available through the Editor. Anyway, since signals can't be exported directly, I'd have to export the signal name, which is more error-prone, and the reason why the issue being discussed exists.

---

This limitation is more related to the fact that the "Advanced" connection options aren't available outside of the scene file, so I might open an issue just for that.

But exporting a signal is definitely a usecase, as I described above.

**Comment by Wiktor102** [comment]
Maybe it's not directly linked to this issue, but it's worth mentioning that `Callable` also has this limitation. A simmilar window in the inspector appears but is not functional.

![image](https://github.com/godotengine/godot/assets/56449542/a058da70-2d3a-4648-9d0b-991a803b6579)

**Comment by Flaschenente** [comment]
I was trying to use a signal export to create a component that would change the icon of a button on a certain signal (e.g. a save button: change to check mark if progress is saved; change to question mark if something was changed). This component could then just be added to any other button that should behave like this.

```gdscript
@export var triggers: Array[Signal]
@export var icon: Texture2D

@onready var parent: Button = get_parent()

func _ready():
	for trigger in triggers:
		trigger.connect(change_icon)
	
func change_icon():
	parent.icon = icon
```

**Comment by GabCoolDude** [comment]
Unless I'm missing something I don't think this is possible. I tried fixing this, and loading is a big problem. To create a working signal, you need an `Object` and a `StringName`, I can easily obtain the node path and the name of the signal from the file, but the problem comes from the fact that when parsing values, no nodes are loaded yet, so it's impossible to get the node and therefore to create the signal.

The same problem would apply when trying to load a `Callable`.

**Comment by rnarma321** [comment]
Not sure if this is still being considered... but would be very nice still, just hit the same problem here people were talking about

**Comment by nalmeth** [comment]
With the addition of ```tween_await``` in 4.7, it would be useful to be able to select ```@export```'s of type Signal. I have something I've written that allows me to build tweens in the editor, and it would be nice to be able to add the ability to ```await``` without having to resort to two exports for the object and name (like what @SwatDoge posted).

**Comment by KoBeWi** [comment]
https://github.com/KoBeWi/Godot-Tween-Suite supports Tween animations with await for specific signals in objects.

**Comment by nalmeth** [comment]
> https://github.com/KoBeWi/Godot-Tween-Suite supports Tween animations with await for specific signals in objects.

By doing exactly what I was saying. You pick the script, then it manually gets the list of signals from it to select. Shouldn't really need to implement this manually for every plugin or project that might want to export signal vars.
